soroban-env-host/src/host.rs

```diff
@@ -1225,6 +1244,12 @@ impl VmCallerCheckedEnv for Host {
 
     // Metering: mostly free or already covered by components (e.g. err_general)
     fn get_invoker_type(&self, _vmcaller: &mut VmCaller<Host>) -> Result<u64, HostError> {
+        #[cfg(test)]
```

Comment:
What's the reason for this change and the similar one in `get_invoking_contract `? I'm hesitant to add test code to non-test functions. We would ideally set up the host the right way so these functions work as expected.

Reply:
I'm open to suggestions, but we need to come up with a straightforward way to inject the values for tests as messing with the frames etc has nothing to do with the test logic.

Reply:
>messing with the frames etc has nothing to do with the test logic.

We should lean into using the system as it is in tests. That's how we learn how the system really behaves. It's also how we learn what is wrong with the design of the current system. If the system is really difficult to use in tests, we might want to reconsider its design down the track, but we should lean into using it as fully and completely as possible.

Reply:
I'm not sure what are the best practices for DI in Rust; what I'd do in C++ requires virtual functions which we don't have here.

Reply:
> We should lean into using the system as it is in tests. 

Sure, in integration/system tests. I'd argue that unit tests should mostly cover the relevant logic (token interface here) and mock the details that aren't relevant to the tests (i.e. how exactly we get the invoker).

Reply:
I defer to @graydon @sisuresh on what the most appropriate approach for the token contract living in the host. But I think we should lean into integration/system tests, because they're less likely to change as we refactor the internals.

Fwiw this is the approach we've taken with supporting contract developers writing tests in the SDK, because the chance of testing the actual behavior of the contract and not fabricated behavior is greater. I believe @jonjove would also encouraging because he'd seen how in stellar-core tests that fiddled with internals became brittle over time.

Reply:
> because they're less likely to change as we refactor the internals.

The point here is to test the 'internals' in a way (the behavior of this specific interface), so the changes would be warranted if that behavior changes for whatever reason. Covering the invocation logic is not necessarily a goal, though I agree it would be nice to have a test that covers that too (without relying on the token's behavior much).

Anyway, I don't think we need to argue here - I agree that the current way is hacky and if we had a less hacky way to imitate the invoker, I'd use it. We could rather merge this temporarily and open an issue to clean this up, or figure out a proper way to do that and change the tests later. I would go with the latter only if it's quick enough, since I think it's better to have partial coverage than no coverage at all. I could also remove only the invoker tests temporarily.

Reply:
So the issue here is that the frames are missing right? We have a similar issue open - https://github.com/stellar/rs-soroban-env/issues/485. I think this should be prioritized since we'll run into this issue more often as we test. 

I do have some questions though - 
1. Can you wrap the `TestToken` methods with `with_frame` calls?
2. I'm not sure if this should be used for the same reason Leigh and I mention above... but with testutils enabled, `call_n` pushes a [frame](https://github.com/stellar/rs-soroban-env/blob/91cf595508d475877f855619ad5baeb5f45a3885/soroban-env-host/src/host.rs#L794) onto the stack.

Reply:
I saw that issue. However what we need to do for the tests is imitate calls by an arbitrary account and by an arbitrary contract (that doesn't have to actually exist), which I believe is not exactly what it is about.

> but with testutils enabled, call_n pushes a [frame](https://github.com/stellar/rs-soroban-env/blob/91cf595508d475877f855619ad5baeb5f45a3885/soroban-env-host/src/host.rs#L794) onto the stack.

It's a bit confusing, but `testutils` are not enabled for the host tests themselves. We could add `test` to the respective conditions, but at that point I'm not sure what are we winning here, given that concern is primarily about using test-only code. I'm also not sure if we cover anything additional (besides testutils... which is not what token tests should cover in the first place). I could try using test frames to reduce the amount of test-only code, but I'm not sure if that really solves the issue of DI/mocking.

Reply:
Using the frames was much simpler than I thought, so I got rid of the test-invoker functions. This comes at a cost of marking `TestContract` frame as test, but it's probably still cleaner. PTAL

Reply:
> This comes at a cost of marking TestContract frame as test, but it's probably still cleaner.

Assuming you mean the following, I think that is fine. Generally we should always accompany `features = "testutils"` with `test` so that all test utilities are available in tests by default.
```rust
#[cfg(any(test, feature = "testutils"))]
TestContract(Hash, Symbol),
```
There are a couple reasons for this:
- If testutils are reasonable for consumers of our crate to use in their tests, they should be reasonable for us to use.
- It prevents us from needing to run tests twice once with testutils and once without, so it reduces build time and test time.

Reply:
Sounds good, I'll keep this in mind for the future changes.